File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/addoredittask/AddOrEditTaskFragment.java
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.addoredittask;

import android.text.Editable;
import android.view.View;
import android.widget.EditText;

import com.zhuinden.simplestackdemoexamplefragments.R;
import com.zhuinden.simplestackdemoexamplefragments.application.Injector;
import com.zhuinden.simplestackdemoexamplefragments.util.BaseFragment;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnTextChanged;
import butterknife.Unbinder;

/**
 * Created by Owner on 2017. 01. 26..
 */

// UNSCOPED!
public class AddOrEditTaskFragment
        extends BaseFragment<AddOrEditTaskFragment, AddOrEditTaskPresenter> {
    private AddOrEditTaskPresenter addOrEditTaskPresenter;

    @OnTextChanged(R.id.add_task_title)
    public void titleChanged(Editable editable) {
        addOrEditTaskPresenter.updateTitle(editable.toString());
    }

    @OnTextChanged(R.id.add_task_description)
    public void descriptionChanged(Editable editable) {
        addOrEditTaskPresenter.updateDescription(editable.toString());
    }

    @BindView(R.id.add_task_title)
    EditText addTaskTitle;

    @BindView(R.id.add_task_description)
    EditText addTaskDescription;

    @Override
    public AddOrEditTaskPresenter getPresenter() {
        return addOrEditTaskPresenter;
    }

    @Override
    public AddOrEditTaskFragment getThis() {
        return this;
    }

    @Override
    protected Unbinder bindViews(View view) {
        return ButterKnife.bind(this, view);
    }

    @Override
    protected void injectSelf() {
        addOrEditTaskPresenter = Injector.get().addOrEditTaskPresenter();
    }

    public void saveTask() {
        addOrEditTaskPresenter.saveTask();
    }

    public void navigateBack() {
        addOrEditTaskPresenter.navigateBack();
    }

    public void setTitle(String title) {
        addTaskTitle.setText(title);
    }

    public void setDescription(String description) {
        addTaskDescription.setText(description);
    }
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/presentation/paths/addoredittask/AddOrEditTaskFragment.kt
package com.zhuinden.simplestackdemoexamplefragments.presentation.paths.addoredittask

import android.app.Activity
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import com.zhuinden.simplestackdemoexamplefragments.util.BaseFragment
import com.zhuinden.simplestackdemoexamplefragments.util.MvpPresenter
import com.zhuinden.simplestackdemoexamplefragments.util.backstackDelegate
import com.zhuinden.simplestackdemoexamplefragments.util.onTextChanged
import kotlinx.android.synthetic.main.path_addoredittask.*

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

// UNSCOPED!
class AddOrEditTaskFragment : BaseFragment<AddOrEditTaskFragment, AddOrEditTaskFragment.Presenter>() {
    companion object {
        const val CONTROLLER_TAG = "AddOrEditTaskPresenter"
    }

    interface Presenter: MvpPresenter<AddOrEditTaskFragment> {
        fun onTitleChanged(title: String)

        fun onDescriptionChanged(description: String)

        fun onSaveButtonClicked()
    }
    
    override val presenter: Presenter by lazy {
        backstackDelegate.lookupService<AddOrEditTaskFragment.Presenter>(CONTROLLER_TAG)
    }

    override fun getThis(): AddOrEditTaskFragment = this

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        textAddTaskTitle.onTextChanged { title ->
            presenter.onTitleChanged(title)
        }

        textAddTaskDescription.onTextChanged { description ->
            presenter.onDescriptionChanged(description)
        }
    }

    fun setTitle(title: String) {
        textAddTaskTitle.setText(title)
    }

    fun setDescription(description: String) {
        textAddTaskDescription.setText(description)
    }

    fun hideKeyboard() {
        (requireActivity().getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager).also { imm ->
            val view = view
            if(view != null) {
                imm.hideSoftInputFromWindow(view.windowToken, 0);
            }
        }
    }

    fun onSaveButtonClicked() {
        presenter.onSaveButtonClicked()
    }
}
